### DISCUSSION TITLE
Package the jvm with the jar

### DISCUSSION THREAD

**Discussion by gdoenlen** [forcedotcom/salesforcedx-vscode#5773] [Ideas]
**Is your feature request related to a problem? Please describe.**
Everyday we have users in the salesforce discord who have problems with installing the jvm and setting the path to it in their settings. This isn't something the end user should be doing.

**Describe the solution you'd like**
The jar for the language server should be packaged as an executable with no dependencies on the user's system.

Options:
* Use jlink to produce a custom runtime image and then package it with jpackage
[jlink](https://docs.oracle.com/javase/9/tools/jlink.htm#JSWOR-GUID-CECAC52B-CFEE-46CB-8166-F17A8E9280E9)
[jpackage](https://openjdk.java.net/jeps/343)

* Use graalvm to make a native binary
[native-image](https://www.graalvm.org/docs/reference-manual/native-image/)

**Comment by lcampos** [comment]
Hi @gdoenlen, thanks for opening this issue. To provide some background, we initially looked at ways in which we could avoid having users download the Java JDK. It turned out that we couldn't ship with all the bundled dependencies because of legal reasons (at the time we only supported Java 8). Now that we support other non-Oracle JDKs we can take a look at this again and see if we can ship everything needed to run the Apex language server jar.

cc @smaddox-sf @vazexqi

**Comment by gdoenlen** [comment]
I completely understand the difficulties around Oracle.  Thanks for the background info.

**Comment by gdoenlen** [comment]
Any updates on this? Seems like you guys support modern jvms now, it should be trivial to use jlink. In the discord server we're having more people who have issues because they have multiple jdk's installed, or path variables and JAVA_HOME not set correctly. There's also been confusion about whether you need to set JRE_HOME (the JRE doesn't exist anymore) or JAVA_HOME.

@lcampos

**Comment by lcampos** [comment]
Hi @gdoenlen , this is still in our plans but it's not the highest priority at the moment. Right now we are focusing on fixing performance issues around deploy/retrieve, running apex tests, and the Apex language server. I think once we're done with those we'll start looking at this one.

Specific to the confusion around the configuration, the only supported settings/variables are the ones documented here [Java Setup](https://developer.salesforce.com/tools/vscode/en/getting-started/java-setup/#site-main-content). I'm not part of that discord server but feel free to pass this along.

cc @smaddox-sf

**Comment by gdoenlen** [comment]
Hi this is my bi-annual check in on this. JDK 17 will be released next month and the current Java 11 runtime that your users are running will be no longer supported by most vendors. This will either leave your users with an insecure runtime or multiple headaches because they have to upgrade their JVM. This can all be solved by packaging the VM with the jar and not placing this responsibility upon your users.

**Comment by gdoenlen** [comment]
Hello again, it's only been 2 years. Here's another user having trouble with this:

![jvm](https://user-images.githubusercontent.com/17297466/161071215-9267c2a0-c492-450a-809d-cc3fd645cd94.png)

**Comment by gdoenlen** [comment]
Hello again, it has been 5 years. Any update? Do you need to hire me to do it for you?
